jib-core/src/main/java/com/google/cloud/tools/jib/filesystem/LockFile.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.filesystem;
+
+import com.google.common.base.Preconditions;
+import java.io.Closeable;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.channels.FileLock;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/** Creates and deletes lock files. */
+public class LockFile implements Closeable {
+
+  private static final ConcurrentHashMap<Path, Lock> lockMap = new ConcurrentHashMap<>();
+
+  private final Path lockFile;
+  private final FileLock fileLock;
+
+  private LockFile(Path lockFile, FileLock fileLock) {
+    this.lockFile = lockFile;
+    this.fileLock = fileLock;
+  }
+
+  /**
+   * Creates a lock file.
+   *
+   * @param lockFile the path of the lock file
+   * @return a new {@link LockFile} that can be released later
+   * @throws IOException if creating the lock file fails
+   */
+  public static LockFile lock(Path lockFile) throws IOException {
+    lockMap.computeIfAbsent(lockFile, key -> new ReentrantLock()).lock();
+    Files.createDirectories(lockFile.getParent());
+    FileLock fileLock = new FileOutputStream(lockFile.toFile()).getChannel().lock();
+    return new LockFile(lockFile, fileLock);
+  }
+
+  /** Releases the lock file. */
+  @Override
+  public void close() {
+    try {
+      fileLock.release();
+
+    } catch (IOException ex) {
+      throw new IllegalStateException("Unable to release lock", ex);
+
+    } finally {
```

Comment:
Sorry, but for the sake of completeness, I think this should further be

```java
  } finally {
    try {
      ....unlock();
      Files.delete(lockFile);
    } catch (IllegalMonitorStateException | IOException ignored) {
    }
  }
```
because if this instance failed to `lock()` at the beginning due to, e.g., interruption, it will throw `IllegalMonitorStateException` when the caller escapes from the try-resource block: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReentrantLock.html#unlock--

Reply:
FTR: I think I was wrong: https://github.com/GoogleContainerTools/jib/pull/1696#discussion_r281347986